monitoring/system_pods.go

```diff
@@ -0,0 +1,249 @@
+		return nil, utils.ConvertError(err)
+	}
+
+	return pods.Items, nil
+}
+
+// verifyPods verifies the pods are in a valid state. Reports a failed probe for
+// each failed pod.
+// https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase
+func (r *systemPodsChecker) verifyPods(pods []corev1.Pod, reporter health.Reporter) {
+	for _, pod := range pods {
+		switch pod.Status.Phase {
+
+		// PodSucceeded indicates that all containers terminated without an error.
+		case corev1.PodSucceeded:
+			continue
+
+		// PodRunning usually indicates the pod is healthy, but the containers
+		// need to be further verified to check if any containers are in a
+		// CrashLoopBackOff state.
+		case corev1.PodRunning:
+			if err := verifyContainers(pod.Status.ContainerStatuses); err != nil {
+				reporter.Add(systemPodsFailureProbe(r.Name(), pod.Namespace, pod.Name, err))
+			}
+
+		// PodPending indicates that some containers are still in a waiting state.
+		// All initContainers and containers need to be verified to ensure that
+		// a container is not stuck in an unhealthy state.
+		case corev1.PodPending:
+			var err error
+
+			// If the pod has already been initialized, skip initContainer
+			// verification and only verify containers.
+			if isInitialized(pod.Status.Conditions) {
+				err = verifyContainers(pod.Status.ContainerStatuses)
+			} else {
+				err = verifyContainers(pod.Status.InitContainerStatuses)
+			}
+			if err != nil {
+				reporter.Add(systemPodsFailureProbe(r.Name(), pod.Namespace, pod.Name, err))
+			}
+
+		// PodFailed indicates that a container terminated with an error.
+		case corev1.PodFailed:
+			err := trace.BadParameter("pod failed: %v", pod.Status.Reason)
+			reporter.Add(systemPodsFailureProbe(r.Name(), pod.Namespace, pod.Name, err))
+
+		// Log any unexpected pod phases and contiune.
+		default:
+			log.WithField("pod", pod.Name).
+				WithField("namespace", pod.Namespace).
+				WithField("phase", pod.Status.Phase).
+				Debug("Pod is in an unknown phase.")
+		}
+	}
+}
+
+// isInitialized returns true if the pod's `Initialized` condition is `True`.
+func isInitialized(conditions []corev1.PodCondition) bool {
+	for _, condition := range conditions {
+		if condition.Type == corev1.PodInitialized && condition.Status == corev1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+// verifyContainers verifies all containers are in a healthy state.
+// Returns an error with state and reason if a container is in an unhealthy state.
+func verifyContainers(containerStatuses []corev1.ContainerStatus) error {
+	for _, status := range containerStatuses {
+		if status.State.Waiting != nil {
+			reason := status.State.Waiting.Reason
+			if reason == errImagePullBackOff || reason == errCrashLoopBackOff || reason == errImagePull {
+				return trace.BadParameter("%v waiting: %v", status.Name, status.State.Waiting.Reason)
+			}
+		}
+		if status.State.Terminated != nil {
+			reason := status.State.Terminated.Reason
```

Comment:
nit: I don't remember if it was discussed but what about exit code != 0 - has this been considered as a failing case as well?

Reply:
Hmm, checking if exit code != 0 probably works better here. Will update.